v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/iowrapper/JdbcIoWrapper.java

```diff
@@ -0,0 +1,194 @@
+    SchemaDiscovery schemaDiscovery =
+        new SchemaDiscoveryImpl(config.dialectAdapter(), config.schemaDiscoveryBackOff());
+    SourceSchema.Builder sourceSchemaBuilder =
+        SourceSchema.builder().setSchemaReference(config.sourceSchemaReference());
+    DataSource dataSource = dataSourceConfiguration.buildDatasource();
+    ImmutableList<String> tables =
+        config.tableConfigs().stream()
+            .map(TableConfig::tableName)
+            .collect(ImmutableList.toImmutableList());
+    ImmutableMap<String, ImmutableMap<String, SourceColumnType>> tableSchemas =
+        schemaDiscovery.discoverTableSchema(dataSource, config.sourceSchemaReference(), tables);
+    tableSchemas.entrySet().stream()
+        .map(
+            tableEntry -> {
+              SourceTableSchema.Builder sourceTableSchemaBuilder =
+                  SourceTableSchema.builder().setTableName(tableEntry.getKey());
+              tableEntry
+                  .getValue()
+                  .entrySet()
+                  .forEach(
+                      colEntry ->
+                          sourceTableSchemaBuilder.addSourceColumnNameToSourceColumnType(
+                              colEntry.getKey(), colEntry.getValue()));
+              return sourceTableSchemaBuilder.build();
+            })
+        .forEach(sourceSchemaBuilder::addTableSchema);
+    return sourceSchemaBuilder.build();
+  }
+
+  private static PTransform<PBegin, PCollection<SourceRow>> getJdbcIO(
+      JdbcIOWrapperConfig config,
+      DataSourceConfiguration dataSourceConfiguration,
+      TableConfig tableConfig,
+      SourceTableSchema sourceTableSchema) {
+    ReadWithPartitions<SourceRow, @UnknownKeyFor @NonNull @Initialized Long> jdbcIO =
+        JdbcIO.<SourceRow>readWithPartitions()
+            .withTable(tableConfig.tableName())
+            .withPartitionColumn(tableConfig.partitionColumns().get(0))
+            .withDataSourceProviderFn(JdbcIO.PoolableDataSourceProvider.of(dataSourceConfiguration))
+            .withRowMapper(
+                new JdbcSourceRowMapper(config.valueMappingsProvider(), sourceTableSchema));
+    if (tableConfig.maxFetchSize() != null) {
+      jdbcIO = jdbcIO.withFetchSize(tableConfig.maxFetchSize());
+    }
+    if (tableConfig.maxPartitions() != null) {
+      jdbcIO = jdbcIO.withNumPartitions(tableConfig.maxPartitions());
+    }
+    return jdbcIO;
+  }
+
+  static DataSourceConfiguration getDataSourceConfiguration(JdbcIOWrapperConfig config) {
+
+    return JdbcIO.DataSourceConfiguration.create(
+            StaticValueProvider.of(config.jdbcDriverClassName()),
+            StaticValueProvider.of(getUrl(config)))
+        .withUsername(config.dbAuth().getUserName())
+        .withPassword(config.dbAuth().getPassword())
+        .withDriverJars(config.jdbcDriverJars())
+        .withMaxConnections(Math.toIntExact(config.maxConnections()));
+  }
+
+  static String getUrl(JdbcIOWrapperConfig config) {
```

Comment:
We should add a todo to unit test this method since I see a few Codecov comments in here.

Reply:
Yup. I am tracking UT in b/339174207
I think only 3 files are not covered in this PR. I will track the report and fix in a followup.